src/clarity-angular/datagrid/providers/filters.ts

```diff
@@ -57,14 +56,21 @@ export class Filters {
     /**
      * Registers a filter, and returns a deregistration function
      */
-    public add(filter: Filter<any>): () => void {
+    public add<F extends Filter<any>>(filter: F): RegisteredFilter<F> {
         let index = this._all.length;
         let subscription = filter.changes.subscribe(() => this._change.next([filter]));
-        this._all.push({filter, subscription});
-        return () => {
+        let hasUnregistered = false;
+        let registered = new RegisteredFilter(filter, () => {
+            if (hasUnregistered) { return; }
             subscription.unsubscribe();
             this._all.splice(index, 1);
-        };
+            if (filter.isActive()) {
+                this._change.next([]);
+            }
+            hasUnregistered = true;
```

Comment:
Could you add a unit test for this `hasUnregistered` check, just to be sure?
Manually register 3 filters, then unregister the middle one twice in a row. Change should only be fired 1 time if the filter was active, and the other two filters should still be present.